library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/impl/QuiltBuiltinResourcePackProfile.java

```diff
@@ -0,0 +1,132 @@
+import net.minecraft.resource.metadata.PackResourceMetadata;
+import net.minecraft.text.LiteralText;
+import net.minecraft.text.Text;
+import net.minecraft.text.TranslatableText;
+import net.minecraft.util.Formatting;
+
+import org.quiltmc.qsl.resource.loader.api.ResourcePackActivationType;
+
+@ApiStatus.Internal
+public final class QuiltBuiltinResourcePackProfile extends ResourcePackProfile {
+	private static final Logger LOGGER = LogUtils.getLogger();
+
+	static QuiltBuiltinResourcePackProfile of(ModNioResourcePack pack) {
+		try {
+			PackResourceMetadata metadata = pack.parseMetadata(PackResourceMetadata.READER);
+			if (metadata == null) {
+				LOGGER.warn("Couldn't find pack meta for pack {}", pack.getName());
+				return null;
+			}
+
+			return new QuiltBuiltinResourcePackProfile(pack, metadata);
+		} catch (IOException e) {
+			LOGGER.warn("Couldn't get pack info for: {}", e.toString());
+			return null;
+		}
+	}
+
+	private QuiltBuiltinResourcePackProfile(ModNioResourcePack pack, PackResourceMetadata metadata) {
+		super(
+				pack.getName(),
+				pack.getDisplayName(),
+				pack.getActivationType() == ResourcePackActivationType.ALWAYS_ENABLED,
+				() -> pack,
+				metadata,
+				pack.type,
+				ResourcePackProfile.InsertionPosition.TOP,
+				new BuiltinResourcePackSource(pack)
+		);
+	}
+
+	@Override
+	public ResourcePackCompatibility getCompatibility() {
+		// This is to ease multi-version mods whose built-in packs actually work across versions.
+		return ResourcePackCompatibility.COMPATIBLE;
+	}
+
+	/**
+	 * Represents a built-in resource pack source.
+	 * Similar to {@link ResourcePackSource#PACK_SOURCE_BUILTIN} but specifies the mod name too.
+	 */
+	public static class BuiltinResourcePackSource implements ResourcePackSource {
+		private static final Text SOURCE_BUILTIN_TEXT = new TranslatableText("pack.source.builtin");
+		private final ModNioResourcePack pack;
+		private final Text text;
+
+		BuiltinResourcePackSource(ModNioResourcePack pack) {
+			this.pack = pack;
+
+			String displayName = pack.getDisplayName().getString();
+			String name = pack.modInfo.getName();
+			Text text = null;
+
+			// We search if the pack mentions the name of the mod from which it's coming from to determine the source text.
+			if (name != null) {
+				if (displayName.contains(name)) {
+					text = SOURCE_BUILTIN_TEXT;
+				}
+			}
+			if (displayName.contains(pack.modInfo.getId())) {
+				text = SOURCE_BUILTIN_TEXT;
+			}
+
+			// If the name doesn't say anything about the source mod, then say it in the source text.
+			// Abbreviations are needed due to GUI constraints on client.
+			if (text == null) {
+				if (name == null || (name.length() > 22 && name.length() > pack.modInfo.getId().length())) {
+					name = pack.modInfo.getId();
+				}
+
+				name = StringUtils.abbreviate(name, 22);
+				this.text = new TranslatableText("options.generic_value", SOURCE_BUILTIN_TEXT, name);
+			} else {
+				this.text = text;
+			}
```

Comment:
This looks like a lot of confusing logic without a clear purpose. If the goal is to make the pack always have the mod id, can you not just trust the mod to have a descriptive pack name? That would also mean that mods won't have to mixin into impl if the mod name is in the description, but isn't exactly the mod id

Reply:
Trusting every mod authors to be honest, or think of things like this?
Kinda hard to be honest.
The logic does not modify the text if the given name is already descriptive enough.

Reply:
If tooltips are an option, adding a tooltip that states the source mod by name/ID is a much better option than this mess of `Text` manipulation.

Reply:
Tooltip has been added to replace this mess thanks to @EnnuiL!